src/Analyzers/CSharp/Analyzers/UseUTF8StringLiteral/UseUTF8StringLiteralDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,157 @@
+                return;
+
+            if (!TryConvertToUTF8String(builder: null, values))
+                return;
+
+            // Only raise the diagnostic on the first token (usually "new")
+            var location = arrayCreationOperation.Syntax.GetFirstToken().GetLocation();
+
+            // Store the original syntax location so the code fix can find the operation again
+            var additionalLocations = ImmutableArray.Create(arrayCreationOperation.Syntax.GetLocation());
+
+            // If this array creation is not an array creation expression, then it must be from
+            // a parameter array, and the Syntax will be the entire invocation. To issue better
+            // diagnostics construct a different location for this case.
+            if (arrayCreationOperation.Syntax is not (ImplicitArrayCreationExpressionSyntax or ArrayCreationExpressionSyntax))
+            {
+                // Issue the diagnostic for all of the parameters that make up the array. We could do just
+                // the first element, but that might be odd seeing: M(1, 2, [|3|], 4, 5)
+                var span = TextSpan.FromBounds(elements.First().Syntax.SpanStart, elements.Last().Syntax.Span.End);
+                location = Location.Create(location.SourceTree!, span);
+            }
+
+            context.ReportDiagnostic(
+                DiagnosticHelper.Create(Descriptor, location, option.Notification.Severity, additionalLocations, properties: null));
+        }
+
+        internal static bool TryConvertToUTF8String(StringBuilder? builder, SegmentedList<byte> values)
+        {
+            // Since we'll only ever need to use up to 4 bytes to check/convert to UTF8
+            // we can just use one array and reuse it. Using an array pool would do the same
+            // thing but with more locks.
+            var array = new byte[4];
+            for (var i = 0; i < values.Count;)
+            {
+                // We only need max 4 elements for a single Rune
+                var count = Math.Min(values.Count - i, 4);
+
+                // Need to copy to a regular array to get a ROS for Rune to process
+                values.CopyTo(i, array, 0, count);
+                var ros = new ReadOnlySpan<byte>(array, 0, count);
+
+                // If we can't decode a rune from the array then it can't be represented as a string
+                if (Rune.DecodeFromUtf8(ros, out var rune, out var bytesConsumed) != System.Buffers.OperationStatus.Done)
+                    return false;
+
+                i += bytesConsumed;
+
+                if (builder is not null)
+                {
+                    if (rune.TryGetEscapeCharacter(out var escapeChar))
+                    {
+                        builder.Append('\\');
+                        builder.Append(escapeChar);
+                    }
+                    else
+                    {
+                        builder.Append(rune.ToString());
```

Comment:
i'm confused.  builder is nullable.  but you're not checking for null.  also, i like bulder being nullable so we don't allocate in the analyzer.  can you make this `?.` instead?

Reply:
This whole thing is inside a null check on line 140